lib/spack/llnl/util/filesystem.py

```diff
@@ -617,6 +617,29 @@ def get_filetype(path_name):
     return output.strip()
 
 
+@system_path_filter
+def is_nonsymlink_exe_with_shebang(path):
+    """
+    Returns whether the path is an executable script with a shebang.
+    Return False when the path is a *symlink* to an executable script.
+    """
+    try:
+        st = os.lstat(path)
+        # Should not be a symlink
+        if stat.S_ISLNK(st.st_mode):
+            return False
+
+        # Should be executable
+        if not st.st_mode & (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH):
```

Comment:
If the stat is already necessary because of the symlink check then this makes sense, but if not for that, any reason not to use `os.access(path, O_X)`?  I suppose the checks are different, but "is this executable by me" seems more like what we want than "is this executable by someone in theory" like the current check, unless this should work for things the user can't run?

Reply:
To answer both comments:
- I didn't consider `os.access` because I copied from the sbang hook (looking at it again it that hook should improve...)
- Since `lstat` was already used, defensively checking it's not a link is free. Possibly this function could be used in other places too.

Reply:
Seems fair, thanks for the explanation @haampie.  It's moderately unfortunate that we still have the extra stats one level up, but this is already so much better I'm not sure I'd push for it here.

Reply:
At the end of the day the gains are minimal, since there's only "double" stat calls in `<pythonp prefix>/bin/`. However, there's a lot of redundant calls to `path_contains_subdirectory` for every file. Maybe a slightly better implementation would be to not force the package to loop over all files, but instead just let spack create a standard view with all symlinks/hardlink/copies -- and then have the package only touch a subset of the files, like `bin/*`.